### PR TITLE
Support for relaxed types

### PR DESCRIPTION
**Pull request by daniel-chambers**
This PR adds support for "relaxed types". "Relaxed types" are types that are otherwise unsupported (ie. we don't know how to make an NDC schema for them (yet)), but instead of being rejected are instead converted into opaque scalar types. To opt into using relaxed types, one must apply the `@allowrelaxedtypes` JSDoc tag to the function that will be using the unsupported types.

Please see the readme for more information.

JIRA: [NDC-389](https://hasurahq.atlassian.net/browse/NDC-389)

## Solution Design
The `ScalarTypeDefinition` and `ObjectTypeDefinition` (in `schema.ts`) now have markers on them that describe whether or not the type is a relaxed type. A relaxed type `ScalarTypeDefinition` represents an unsupported type that has been munged into a opaque scalar, and a relaxed type `ObjectTypeDefinition` is simply an object type that uses a relaxed type somewhere in its properties.

All the existing `TypeDefinition` subtypes got renamed to `TypeReference` (eg `ArrayTypeDefinition` -> `ArrayTypeReference`), because they aren't really "definitions", they're actually just references to types defined elsewhere.

`inference.ts`'s `deriveSchemaTypeForTsType` function has been refactored significantly to extract out all the previously rejected types that are now allowed in relaxed type mode into separate functions which validate them and emit scalar types.

For unsupported types that are composite types (ie are made up of other types, for example `Record<string, number>`), they will only be allowed to be used as a relaxed type if the types they compose are allowed as supported or relaxed types. This prevents users from using relaxed types to allow nonsensical types such as `Record<string, void>`.

New unit tests have been added to cover all the permutations of relaxed types in `test/inference/relaxed-types`.

[NDC-389]: https://hasurahq.atlassian.net/browse/NDC-389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -147,6 +147,41 @@ These types are unsupported as function parameter types or return types for func
...
+    return "Error: record was empty";
+
+  for (const [key, value] of entries) {
+    if (value === "")
+      emptyKeys.push(key);
+  }
+
+  return { emptyKeys };
+}
+```
```
Comment (sordina) [inline README.md:183]: Nice example 👍

File: ndc-lambda-sdk/src/inference.ts [hunk 2]
```diff
@@ -226,154 +237,209 @@ function getParallelDegreeFromJsDoc(functionSymbol: ts.Symbol, functionIsReadonl
...
-  if (context.typeChecker.getIndexInfosOfType(tsType).length > 0) {
-    return new Err([`Types with index signatures are not supported, but one was encountered in ${typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)})`]);
+  // We don't know how to deal with this type, so reject it with a generic error
+  return new Err([`Unable to derive an NDC type for ${schema.typePathToString(typePath)} (type: ${context.typeChecker.typeToString(tsType)}).`]);
+}
+
+function deriveRelaxedTypeOrError(typeName: string, typePath: schema.TypePathSegment[], mkError: () => string, allowRelaxedTypes: boolean, context: TypeDerivationContext): Result<schema.TypeReference, string[]> {
+  if (allowRelaxedTypes === false) {
+    return new Err([mkError()]);
   }
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:299]: Do you already know that this is a relaxed type when you call it?

File: ndc-lambda-sdk/src/inference.ts [hunk 3]
```diff
@@ -226,154 +237,209 @@ function getParallelDegreeFromJsDoc(functionSymbol: ts.Symbol, functionIsReadonl
...
     return new Ok({ type: "named", kind: "scalar", name: schema.BuiltInScalarTypeName.BigInt, literalValue: literalValue });
   }
   if (isDateType(tsType)) {
-    context.scalarTypeDefinitions[schema.BuiltInScalarTypeName.DateTime] = {};
+    context.scalarTypeDefinitions[schema.BuiltInScalarTypeName.DateTime] = { type: "built-in" };
     return new Ok({ type: "named", kind: "scalar", name: schema.BuiltInScalarTypeName.DateTime });
   }
   if (isJSONValueType(tsType, context.ndcLambdaSdkModule)) {
-    context.scalarTypeDefinitions[schema.BuiltInScalarTypeName.JSON] = {};
+    context.scalarTypeDefinitions[schema.BuiltInScalarTypeName.JSON] = { type: "built-in" };
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:442]: Should we have a constant for `{ type: "built-in" }`?

Comment (sordina) [reply ndc-lambda-sdk/src/inference.ts:442]: In fact, something about this seems a little fragile, maybe a helper for the `context.scalarTypeDefinitions[schema.BuiltInScalarTypeName.JSON] = { type: "built-in" }` pattern?

File: ndc-lambda-sdk/src/inference.ts [hunk 4]
```diff
@@ -388,7 +454,7 @@ function isFunctionType(tsType: ts.Type): boolean {
   return tsType.getCallSignatures().length > 0 || tsType.getConstructSignatures().length > 0;
 }
 
-function isMapType(tsType: ts.Type): boolean {
+function isMapType(tsType: ts.Type): tsType is ts.TypeReference {
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:457]: Nice.

File: ndc-lambda-sdk/src/inference.ts [hunk 5]
```diff
@@ -422,38 +488,69 @@ function isJSONValueType(tsType: ts.Type, ndcLambdaSdkModule: ts.ResolvedModuleF
...
+      }
+
+      context.objectTypeDefinitions[info.generatedTypeName] = { properties: propertyResults, description: info.description, isRelaxedType }
+      return new Ok({ type: 'named', name: info.generatedTypeName, kind: "object" } as const)
+    })
+    .onErr(_err => {
+      delete context.objectTypeDefinitions[info.generatedTypeName]; // Remove the recursion short-circuit to allow other functions to try making this type again
     });
+  }
+}
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:533]: Quite a lot of changes, so a bit hard to tell if anything is off. I'll focus on the test-cases and see if that all looks to preserve the behaviour.

File: ndc-lambda-sdk/src/inference.ts [hunk 6]
```diff
@@ -422,38 +488,69 @@ function isJSONValueType(tsType: ts.Type, ndcLambdaSdkModule: ts.ResolvedModuleF
...
+        case "scalar":
+          const scalarTypeDef = context.scalarTypeDefinitions[typeReference.name] ?? throwError(`Unable to find object type '${typeReference.name}', which should have already been generated`);
+          return scalarTypeDef.type === "relaxed-type";
+        default:
+          return unreachable(typeReference["kind"]);
+      }
+    default:
+      return unreachable(typeReference["type"]);
   }
 }
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:555]: 👍

File: ndc-lambda-sdk/src/schema.ts [hunk 7]
```diff
@@ -36,95 +36,105 @@ export type ObjectTypeDefinitions = {
...
-  type: TypeDefinition,
+  type: TypeReference,
 }
 
 export type ScalarTypeDefinitions = {
   [scalarTypeName: string]: ScalarTypeDefinition
 }
 
-export type ScalarTypeDefinition = Record<string, never> // Empty object, for now
+export type ScalarTypeDefinition = BuiltInScalarTypeDefinition | RelaxedScalarTypeDefinition
```
Comment (sordina) [inline ndc-lambda-sdk/src/schema.ts:53]: 👍

File: ndc-lambda-sdk/src/schema.ts [hunk 8]
```diff
@@ -36,95 +36,105 @@ export type ObjectTypeDefinitions = {
...
 
-export type TypeDefinition = ArrayTypeDefinition | NullableTypeDefinition | NamedTypeDefinition
+export type BuiltInScalarTypeDefinition = {
+  type: "built-in",
+}
+
+export type RelaxedScalarTypeDefinition = {
+  type: "relaxed-type"
+  usedIn: TypePathSegment[][]
```
Comment (sordina) [inline ndc-lambda-sdk/src/schema.ts:61]: Nice.

File: ndc-lambda-sdk/src/schema.ts [hunk 9]
```diff
@@ -190,14 +200,40 @@ export class JSONValue {
...
 }
 
+export type TypePathSegment =
+  { segmentType: "FunctionParameter", functionName: string, parameterName: string }
+  | { segmentType: "FunctionReturn", functionName: string }
+  | { segmentType: "ObjectProperty", typeName: string, propertyName: string }
+  | { segmentType: "Array" }
+  | { segmentType: "TypeParameter", typeName: string, index: number }
+  | { segmentType: "IndexSignature", typeName: string, sigIndex: number, component: "key" | "value" }
+  | { segmentType: "UnionMember", typeName: string, memberIndex: number }
```
Comment (sordina) [inline ndc-lambda-sdk/src/schema.ts:210]: Clean 👍

File: ndc-lambda-sdk/src/schema.ts [hunk 10]
```diff
@@ -190,14 +200,40 @@ export class JSONValue {
...
+
+export function typePathToString(segments: TypePathSegment[]): string {
+  return segments.map(typePathSegmentToString).join(", ")
+}
```
Comment (sordina) [inline ndc-lambda-sdk/src/schema.ts:214]: Should this be '.' style? Otherwise, show as JSON.stringify(foo.map...)

File: ndc-lambda-sdk/src/schema.ts [hunk 11]
```diff
@@ -268,10 +304,29 @@ export function printSchemaListing(functionNdcKind: FunctionNdcKind, functionDef
...
+export function printRelaxedTypesWarning(functionsSchema: FunctionsSchema) {
+  const relaxedTypes = Object.entries(functionsSchema.scalarTypes).flatMap<[string, RelaxedScalarTypeDefinition]>(([scalarTypeName, definition]) => {
+    return definition.type === "relaxed-type"
+      ? [[scalarTypeName, definition]]
+      : []
+  });
+  if (relaxedTypes.length > 0) {
+    console.error("The following unsupported types have been defined as custom scalar types with no runtime validation ('relaxed types').")
+    console.error("Use of relaxed types may cause runtime issues if invalid values are passed for these types.")
+    console.error("It is recommended that you replace them with equivalent supported types so that your schema can be properly defined and validated, then remove the @allowrelaxedtypes tag from the functions involved.")
```
Comment (sordina) [inline ndc-lambda-sdk/src/schema.ts:316]: Can't miss that!

File: ndc-lambda-sdk/test/execution/execute-query.test.ts [hunk 12]
```diff
@@ -289,7 +289,7 @@ describe("execute query", function() {
       },
       objectTypes: {},
       scalarTypes: {
-        "String": {}
+        "String": { type: "built-in" }
```
Comment (sordina) [inline ndc-lambda-sdk/test/execution/execute-query.test.ts:292]: Yeah let's have a constant for sure.

File: ndc-lambda-sdk/test/inference/relaxed-types/any.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+/** @allowrelaxedtypes */
+export function anyFunction(test: any): any {
+  return "wow";
+}
```
Comment (sordina) [inline ndc-lambda-sdk/test/inference/relaxed-types/any.ts:4]: wow (in Owen Wilson's voice)

File: ndc-lambda-sdk/test/inference/relaxed-types/index-signature-types-invalid.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+/** @allowrelaxedtypes */
+export function indexSignatureTypesFunction(
+  recordType: Record<string, void>,
+  inlineSig: { [x: string]: undefined },
+  objWithSig: ObjectWithIndexSignature,
+  objWithPropsAndSig: ObjectWithPropsAndIndexSignature,
+  intersectionSigs: IntersectionOfMultipleIndexSigs
+): string {
+  return ""
+}
```
Comment (sordina) [inline ndc-lambda-sdk/test/inference/relaxed-types/index-signature-types-invalid.ts:28]: Nice!

File: ndc-lambda-sdk/test/inference/relaxed-types/index-signature-types-valid.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+/** @allowrelaxedtypes */
+export function indexSignatureTypesFunction(
+  recordType: Record<string, string>,
+  inlineSig: { [x: string]: string },
+  objWithSig: ObjectWithIndexSignature,
+  objWithPropsAndSig: ObjectWithPropsAndIndexSignature,
+  intersectionSigs: IntersectionOfMultipleIndexSigs
+): string {
+  return ""
+}
```
Comment (sordina) [inline ndc-lambda-sdk/test/inference/relaxed-types/index-signature-types-valid.ts:28]: 👍

File: ndc-lambda-sdk/test/inference/relaxed-types/relaxed-types.test.ts
Comment (sordina) [inline ndc-lambda-sdk/test/inference/relaxed-types/relaxed-types.test.ts:1]: Awesome.

File: ndc-lambda-sdk/test/inference/relaxed-types/tuple-types-invalid.ts
Comment (sordina) [inline ndc-lambda-sdk/test/inference/relaxed-types/tuple-types-invalid.ts:1]: I am curious if some edge-case invalid types will end up having valid uses. Maybe incorrectly specified type interfaces perhaps...

**Review by sordina — APPROVED** [review]
Looks amazing @daniel-chambers 

I've added a few questions and suggestions, but the excellent test-coverage and clear preservation of existing test behaviour give a lot of confidence in this change.

File: ndc-lambda-sdk/src/inference.ts (hunk 2 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/inference.ts:299]: Yes, it's used by the other functions that identify the appropriate types that can be turned into relaxed types (eg `unknown` or `any`).

File: ndc-lambda-sdk/src/inference.ts (hunk 3 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/inference.ts:442]: I dunno, extracting a function for one line of code doesn't really seem worth it, especially when that line is doing something straightforward and obvious. Creating built-in scalar types is also only done in this function (this is _the_ scalar types function), so it's not like it'd get re-used in multiple disparate locations, which would be a a good reason to extract it into a function.

File: ndc-lambda-sdk/src/schema.ts (hunk 11 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/schema.ts:316]: 💣 💣 💣 😂

File: ndc-lambda-sdk/test/execution/execute-query.test.ts (hunk 12 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/test/execution/execute-query.test.ts:292]: Seems hardly worth it, given it's not a magical value (obvious what it is), and that the most likely predicted change to this (adding a property to that object) would require changing in all locations anyway (to add the specific value of that property) and likely the removal of the constant.

I'm not gonna die on this hill though, so if you feel strongly about it, I'll extract all instances to a constant. (Keep in mind `const builtIn = { type: "built-in" }` is not a true constant, it's a potentially mutable shared instance...)

File: ndc-lambda-sdk/test/inference/relaxed-types/tuple-types-invalid.ts
Comment (daniel-chambers) [reply ndc-lambda-sdk/test/inference/relaxed-types/tuple-types-invalid.ts:1]: What an example of an "incorrectly specified type interfaces"? If there's an edge case you're thinking of I'd like to cover it explicitly 🙂

There's so many edge cases and complexity in TS that the only way we're going to be able to have confidence in this code is if we capture and test every (unique) case we can think of, that's why I've gone hard on covering cases in the tests.

File: ndc-lambda-sdk/src/schema.ts (hunk 10 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/schema.ts:214]: I don't think so, because this uses the below function to form sentences for error messages, not syntax (eg "function "test" parameter "param", array type, type "MyObject" property "prop")

File: ndc-lambda-sdk/test/execution/execute-query.test.ts (hunk 12 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/test/execution/execute-query.test.ts:292]: Related comment on ndc spec update PR: https://github.com/hasura/ndc-nodejs-lambda/pull/8#discussion_r1491861932

File: ndc-lambda-sdk/src/inference.ts (hunk 2 above)
Comment (sordina) [reply ndc-lambda-sdk/src/inference.ts:299]: Ah I see. I feel like it's a bit of a smell to pass in a bool like this, since if you already know, then just don't call the function, but I don't mind if you want to keep it this way.

File: ndc-lambda-sdk/src/inference.ts (hunk 3 above)
Comment (sordina) [reply ndc-lambda-sdk/src/inference.ts:442]: What about a constant for type: built-in though?

File: ndc-lambda-sdk/src/inference.ts (hunk 2 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/inference.ts:299]: Oh sorry, I may not have explained properly. This is _always_ called in the functions that identify types that may be relaxed types. This function is responsible for bailing out if relaxed types is disallowed.

For example, see the `deriveSchemaTypeIfTsTupleType`. Yeah, there's an `allowRelaxedTypes` check in there, but it's that's really just preventing un-needed computation if relaxed types is disabled (Haskell's laziness would have been nice here 🥲).

Keeping the bailing in this function means the error message is consistent across all relaxed types.

File: ndc-lambda-sdk/src/schema.ts (hunk 10 above)
Comment (sordina) [reply ndc-lambda-sdk/src/schema.ts:214]: Ah I see. Sure that's ok then.

File: ndc-lambda-sdk/test/execution/execute-query.test.ts (hunk 12 above)
Comment (sordina) [reply ndc-lambda-sdk/test/execution/execute-query.test.ts:292]: Ok well feel free to keep it if you prefer.

File: ndc-lambda-sdk/src/inference.ts (hunk 3 above)
Comment (daniel-chambers) [reply ndc-lambda-sdk/src/inference.ts:442]: Will leave as in as discussed here: https://github.com/hasura/ndc-nodejs-lambda/pull/10#discussion_r1491929573
